Libraries/plist/Tools/PlistBuddy.cpp

```diff
@@ -0,0 +1,572 @@
+
+public:
+    std::string command() const
+    { return _command; }
+
+public:
+    std::string input() const
+    { return _input; }
+
+private:
+    friend class libutil::Options;
+    std::pair<bool, std::string>
+    parseArgument(std::vector<std::string> const &args, std::vector<std::string>::const_iterator *it);
+
+};
+
+Options::
+Options() :
+    _help   (false),
+    _xml    (false),
+    _command("")
+{
+}
+
+Options::
+~Options()
+{
+}
+
+std::pair<bool, std::string> Options::
+parseArgument(std::vector<std::string> const &args, std::vector<std::string>::const_iterator *it)
+{
+    std::string const &arg = **it;
+
+    if (arg == "-h") {
+        return libutil::Options::MarkBool(&_help, arg, it);
+    } else if (arg == "-x") {
+        return libutil::Options::MarkBool(&_xml, arg, it);
+    } else if (arg == "-c") {
+        return libutil::Options::NextString(&_command, args, it);
+    } else if (!arg.empty() && arg[0] != '-') {
+        _input = arg;
+        return std::make_pair(true, std::string());
+    } else {
+        return std::make_pair(false, "unknown argument " + arg);
+    }
+}
+
+plist::ObjectType parseType(std::string &typeString) {
+    if (typeString == "string") {
+        return plist::ObjectType::String;
+    } else if (typeString == "dictionary") {
+        return plist::ObjectType::Dictionary;
+    } else if (typeString == "array") {
+        return plist::ObjectType::Array;
+    } else if (typeString == "bool") {
+        return plist::ObjectType::Boolean;
+    } else if (typeString == "real") {
+        return plist::ObjectType::Real;
+    } else if (typeString == "integer") {
+        return plist::ObjectType::Integer;
+    } else if (typeString == "date") {
+        return plist::ObjectType::Date;
+    } else if (typeString == "data") {
+        return plist::ObjectType::Data;
+    } else {
+        return plist::ObjectType::None;
+    }
+}
+
+static int
+Help(std::string const &error = std::string())
+{
+    if (!error.empty()) {
+        fprintf(stderr, "error: %s\n", error.c_str());
+        fprintf(stderr, "\n");
+    }
+
+    fprintf(stderr, "usage: PlistBuddy [options] <file.plist>\n");
+
+#define INDENT "  "
+    fprintf(stderr, "\noptions:\n");
+    fprintf(stderr, INDENT "-c \"<command>\" command to execute, otherwise run in interactive mode\n");
+    fprintf(stderr, INDENT "-x output will be in xml plist format\n");
+    fprintf(stderr, INDENT "-h print help including commands\n");
+#undef INDENT
+
+    return (error.empty() ? 0 : -1);
+}
+
+static void parseCommandKeyPathString(const std::string &keyPathString, std::queue<std::string> &keyPath) {
+    auto loc = keyPathString.find(':');
+    if (loc == std::string::npos) {
+        if (keyPathString.size() > 0) {
+            keyPath.push(keyPathString);
+        }
+    } else {
+        if (loc > 0) {
+            keyPath.push(keyPathString.substr(0, loc));
+        }
+        parseCommandKeyPathString(keyPathString.substr(loc + 1), keyPath);
```

Comment:
Streams are an unfortunate part of the C++ stdlib, full of operator overloading and mutable references usually avoided in xcbuild. I also avoided `std::getline()` specifically for both a misleadingly name and mutable references. It would be nice if there was a `std::split()` of some kind.